### PR TITLE
Make a light weight version of DEBUG OBJECT, add FAST option

### PR DESCRIPTION
**Pull request by enjoy-binbin**
Adding FAST option to DEBUG OBJECT command.

The light version only shows the light weight infomation,
which mostly O(1). The pre-existing version that show more
stats such as serializedlength sometimes is time consuming.

This should allow looking into debug stats (the key expired
but not deleted), even on huge object, on which we're afraid
to run the command for fear of causing a server freeze.

Like 3ca451c46fed894bf49e7561fa0282d2583f1c06.

### CODE REVIEW

**Review by zuiderkwast — APPROVED** [review]
This is not backward compatible, but we don't need backward compatibility for DEBUG?

**Comment by enjoy-binbin** [comment]
> This is not backward compatible, but we don't need backward compatibility for DEBUG?

yes, it is for debug, i don't think we need backward compatibility for it.

**Comment by madolson** [comment]
> This is not backward compatible, but we don't need backward compatibility for DEBUG?

Why not make it backward compatible though? I found a few examples of AWS using it internally for testing that would fail on rdbSavedObjectLen(). Those are easy to fix, but maybe someone else is using that as well. We could easily make it backwards compatible and do `DEBUG OBJECT FAST` no?

**Comment by enjoy-binbin** [comment]
> Why not make it backward compatible though? I found a few examples of AWS using it internally for testing that would fail on rdbSavedObjectLen(). Those are easy to fix, but maybe someone else is using that as well. We could easily make it backwards compatible and do DEBUG OBJECT FAST no?

the default behavior is a bit dangerous, like act on a large lists, but i guess the `DEBUG` is already dangerous in usually, `DEBUG OBJECT FAST` looks good to me, i will change it

**Review by madolson — APPROVED** [review]
Yeah, this seems great to me.

File: src/debug.c [hunk 1]
```diff
@@ -639,22 +644,25 @@ void debugCommand(client *c) {
...
-                             "lru:%d lru_seconds_idle:%llu%s",
-                             (void *)val, val->refcount, strenc, rdbSavedObjectLen(val, c->argv[2], c->db->id),
-                             val->lru, estimateObjectIdleTime(val) / 1000, extra);
+        sds s = sdsempty();
+        s = sdscatprintf(s, "Value at:%p refcount:%d encoding:%s", (void *)val, val->refcount, strenc);
+        if (!fast) s = sdscatprintf(s, " serializedlength:%zu", rdbSavedObjectLen(val, c->argv[2], c->db->id));
+        s = sdscatprintf(s, " lru:%d lru_seconds_idle:%llu", val->lru, estimateObjectIdleTime(val) / 1000);
+        s = sdscatprintf(s, "%s", extra);
+        addReplyStatusLength(c, s, sdslen(s));
+        sdsfree(s);
```
Comment (zuiderkwast) [inline src/debug.c:665]: Same logic, but avoid many small sdscatprintf and maybe smaller diff like this?

Hm... Maybe your code is better. It is shorter. :)

```suggestion
        sds s = sdsempty();
        if (!fast) s = sdscatprintf(s, "serializedlength:%zu ", rdbSavedObjectLen(val, c->argv[2], c->db->id));
        addReplyStatusFormat(c,
                             "Value at:%p refcount:%d "
                             "encoding:%s %s"
                             "lru:%d lru_seconds_idle:%llu%s",
                             (void *)val, val->refcount, strenc, s,
                             val->lru, estimateObjectIdleTime(val) / 1000, extra);
        sdsfree(s);
```

**Review by zuiderkwast — APPROVED** [review]

File: src/debug.c (hunk 1 above)
Comment (enjoy-binbin) [reply src/debug.c:665]: well, your diff is smaller, mine seems to be more readable, i prefer mine, so in the future we can just keep appending new fields line by line, like this LFU one #479 can be easy to add

**Pull request merged**